File: backend/sdk/tools/gameplay_tools/onboarding_tools.py
"""
Onboarding tools for TRPG world initialization.

This module defines MCP tools that Onboarding_Manager uses during the onboarding phase:
- draft_world: Lightweight world draft (genre, theme, lore summary) to unblock sub-agents
- persist_world: Comprehensive persistence (full lore + stats + player state)
- complete: Transition world from 'onboarding' to 'active' phase

The onboarding flow:
1. Onboarding_Manager interviews player and gathers preferences
2. Onboarding_Manager calls draft_world with genre, theme, lore summary
3. Sub-agents run in background with draft context (location_designer, character_designer, item_designer)
4. Onboarding_Manager calls persist_world with full lore + stat system (overwrites draft)
5. Onboarding_Manager calls complete to finalize phase transition

Sub-agent tool mappings (persist tools by MCP server):
- Onboarding MCP server:
  - draft_world: Used by Onboarding_Manager directly
  - persist_world: Used by Onboarding_Manager directly
- Subagents MCP server (shared by Action Manager and Onboarding Manager):
  - item_designer: Uses mcp__subagents__persist_item
  - character_designer: Uses mcp__subagents__persist_character_design
  - location_designer: Uses mcp__subagents__persist_location_design
- Action Manager direct tools:
  - change_stat: Used directly by Action Manager

Uses WorldService for filesystem-primary storage.
"""

import logging
import uuid
from datetime import datetime
from typing import Any

from claude_agent_sdk import tool
from services.world_service import WorldService
from sqlalchemy.ext.asyncio import AsyncSession

from sdk.config.onboarding_inputs import CompleteOnboardingInput, DraftWorldInput, PersistWorldInput
from sdk.loaders import get_tool_description, get_tool_response, is_tool_enabled
from sdk.tools.context import ToolContext

logger = logging.getLogger("OnboardingTools")


# =============================================================================
# Tool Name Constants
# =============================================================================

# Tool names for each sub-agent type (persist tools)
# Sub-agents use the shared "subagents" MCP server tools (not action_manager)
# This server is available to both Action Manager and Onboarding Manager
SUBAGENT_TOOL_NAMES = {
    "item_designer": "mcp__subagents__persist_item",
    "character_designer": "mcp__subagents__persist_character_design",
    "location_designer": "mcp__subagents__persist_location_design",
}


async def _add_world_agents_to_initial_location(
    db: AsyncSession,
    world_id: int,
    world_name: str,
    initial_location_name: str,
) -> int:
    """
    Add all world-specific agents to the initial location's room.

    This is called after world seed generation to place NPCs created during
    onboarding into the starting location.

    Args:
        db: Database session
        world_id: World ID
        world_name: World name for filesystem lookup
        initial_location_name: Name of the initial location

    Returns:
        Number of agents added to the location
    """
    import crud
    from services.agent_filesystem_service import AgentFilesystemService

    # Get list of agent names from world's agents folder
    world_agent_names = AgentFilesystemService.list_world_agents(world_name)
    if not world_agent_names:
        logger.info("No world agents to add to initial location")
        return 0

    # Get the initial location from DB
    initial_location = await crud.get_location_by_name(db, world_id, initial_location_name)
    if not initial_location:
        logger.warning(f"Initial location '{initial_location_name}' not found in database")
        return 0

    added_count = 0
    for agent_name in world_agent_names:
        try:
            # Get agent from database
            agent = await crud.get_agent_by_name(db, agent_name)
            if agent:
                # Add to initial location
                await crud.add_character_to_location(db, agent.id, initial_location.id)
                logger.info(f"✨ Added '{agent_name}' to initial location '{initial_location_name}'")
                added_count += 1
            else:
                logger.warning(f"Agent '{agent_name}' not found in database")
        except Exception as e:
            logger.warning(f"Failed to add agent '{agent_name}' to location: {e}")

    logger.info(f"Added {added_count} agents to initial location")
    return added_count


def generate_default_world_name() -> str:
    """
    Generate a default world name if one is not provided.

    Uses timestamp and short UUID to create a unique name.

    Returns:
        Generated world name (e.g., 'world_20251206_a1b2c3')
    """
    timestamp = datetime.now().strftime("%Y%m%d")
    short_uuid = uuid.uuid4().hex[:6]
    return f"world_{timestamp}_{short_uuid}"


def create_onboarding_tools(ctx: ToolContext) -> list:
    """
    Create onboarding tools (draft_world, persist_world, complete) with filesystem persistence.

    Args:
        ctx: Unified tool context containing agent info and dependencies

    Returns:
        List of onboarding tool functions configured with agent name
    """
    tools = []

    # ==========================================================================
    # draft_world tool - Lightweight world draft to unblock sub-agents
    # Called FIRST, before sub-agents start
    # ==========================================================================
    tools.append(_create_draft_world_tool(ctx))

    # ==========================================================================
    # persist_world tool - Comprehensive persistence (full lore + stats)
    # Called AFTER sub-agents have started with draft context
    # ==========================================================================
    tools.append(_create_persist_world_tool(ctx))

    # ==========================================================================
    # Complete tool - Lightweight phase transition
    # Called LAST after draft_world, persist_world, and sub-agents
    # ==========================================================================
    if is_tool_enabled("complete"):
        complete_description = get_tool_description("complete", agent_name=ctx.agent_name, group_name=ctx.group_name)

        @tool("complete", complete_description, CompleteOnboardingInput.model_json_schema())
        async def complete_tool(args: dict[str, Any]):
            """
            Finalize world transition from 'onboarding' to 'active' phase.

            This tool:
            1. Sets the player name in world config
            2. Sets pending_phase to 'active' for phase transition
            3. Captures initial state (stats, inventory, location) for reset functionality

            Call this AFTER:
            - draft_world (genre, theme, lore summary)
            - Sub-agents (location_designer, character_designer, item_designer)
            - persist_world (full lore, stats)
            """
            from services.player_service import PlayerService
            from services.world_reset_service import WorldResetService

            validated = CompleteOnboardingInput(**args)
            player_name = validated.player_name

            logger.info(f"World completion requested by {ctx.agent_name}")
            logger.info(f"Player: {player_name}")

            effective_world_name = ctx.world_name
            if not effective_world_name:
                effective_world_name = generate_default_world_name()
                logger.info(f"Generated default world name: {effective_world_name}")

            try:
                # Ensure world directory exists
                config = WorldService.ensure_world_exists(effective_world_name)

                # Set player name and transition phase
                config.user_name = player_name
                config.pending_phase = "active"
                WorldService.save_world_config(effective_world_name, config)

                logger.info(f"✅ World '{effective_world_name}' pending phase set to 'active'")

                # Get player state for initial state capture and location
                player_state = PlayerService.load_player_state(effective_world_name)
                initial_location_name = player_state.current_location if player_state else None

                # Capture initial state for reset functionality
                # This includes inventory created by item_designer during onboarding
                if player_state:
                    initial_state = WorldResetService.create_initial_state_snapshot(
                        starting_location=player_state.current_location or "",
                        initial_stats=player_state.stats,
                        initial_inventory=player_state.inventory,
                    )
                    WorldResetService.save_initial_state(effective_world_name, initial_state)
                    logger.info(
                        f"Captured initial state: {len(player_state.stats)} stats, "
                        f"{len(player_state.inventory)} items"
                    )

                # Add any world agents (NPCs created during onboarding) to the initial location
                if ctx.db is not None and ctx.world_id is not None and initial_location_name:
                    try:
                        await _add_world_agents_to_initial_location(
                            db=ctx.db,
                            world_id=ctx.world_id,
                            world_name=effective_world_name,
                            initial_location_name=initial_location_name,
                        )
                    except Exception as agent_err:
                        logger.warning(f"Failed to add agents to initial location: {agent_err}")

            except Exception as e:
                logger.error(f"Failed to complete onboarding: {e}", exc_info=True)
                return {
                    "content": [{"type": "text", "text": f"Error completing onboarding: {e}"}],
                    "is_error": True,
                }

            # Get response template from config
            response_text = get_tool_response(
                "complete",
                group_name=ctx.group_name,
                player_name=player_name,
            )

            return {"content": [{"type": "text", "text": response_text}]}

        tools.append(complete_tool)

    # ==========================================================================
    # NOTE: add_character tool has been removed from onboarding.
    # Character creation during onboarding should use Task tool to invoke
    # character_designer sub-agent, which calls persist_character_design directly.
    # ==========================================================================

    return tools


# =============================================================================
# Draft World Tool (Lightweight - unblocks sub-agents)
# =============================================================================


def _create_draft_world_tool(ctx: ToolContext):
    """
    Create the draft_world tool for Onboarding_Manager.

    This tool creates a lightweight world draft with genre, theme, and a
    one-paragraph lore summary. Sub-agents can start immediately with this
    context while OM refines the full lore.

    Args:
        ctx: Tool context with world info

    Returns:
        Tool function for drafting world
    """

    @tool(
        "draft_world",
        "Create a lightweight world draft with genre, theme, and lore summary. "
        "Call this FIRST to unblock sub-agents. They will use this context to "
        "create thematically consistent content while you refine the full lore.",
        DraftWorldInput.model_json_schema(),
    )
    async def draft_world(args: dict[str, Any]):
        """Create lightweight world draft to unblock sub-agents."""
        try:
            validated = DraftWorldInput.model_validate(args)
            world_name = ctx.require_world_name()

            # Ensure world directory exists
            config = WorldService.ensure_world_exists(world_name)

            # Save genre and theme to world config
            config.genre = validated.genre
            config.theme = validated.theme
            WorldService.save_world_config(world_name, config)

            # Save lore summary as initial lore.md
            lore_content = f"# World Lore\n\n{validated.lore_summary}"
            WorldService.save_lore(world_name, lore_content)

            logger.info(f"✅ World draft created for '{world_name}'")
            logger.info(f"Genre: {validated.genre}, Theme: {validated.theme}")

            return {
                "content": [
                    {
                        "type": "text",
                        "text": f"World draft created. Genre: {validated.genre}, Theme: {validated.theme}. "
                        f"Sub-agents can now start with this context.",
                    }
                ],
            }

        except Exception as e:
            logger.error(f"Failed to create world draft: {e}", exc_info=True)
            return {
                "content": [{"type": "text", "text": f"Error creating world draft: {e}"}],
                "is_error": True,
            }

    return draft_world


# =============================================================================
# Persist World Tool (Comprehensive - consolidates everything)
# =============================================================================


def _create_persist_world_tool(ctx: ToolContext):
    """
    Create the persist_world tool for Onboarding_Manager.

    This tool consolidates full lore with stat system and player state.
    Called AFTER sub-agents have started with draft context.
    Overwrites the draft lore with the full version.

    Args:
        ctx: Tool context with world info

    Returns:
        Tool function for persisting world
    """

    @tool(
        "persist_world",
        "Persist comprehensive world data: full lore (overwrites draft) + stat system + player state. "
        "Call this AFTER sub-agents have started with draft_world context.",
        PersistWorldInput.model_json_schema(),
    )
    async def persist_world(args: dict[str, Any]):
        """Persist comprehensive world data (full lore + stats)."""
        from services.player_service import PlayerService

        try:
            validated = PersistWorldInput.model_validate(args)
            world_name = ctx.require_world_name()

            # Ensure world directory exists
            WorldService.ensure_world_exists(world_name)

            # Save stat definitions
            stat_definitions = {
                "stats": [stat.model_dump() for stat in validated.stat_system.stats],
                "derived": validated.stat_system.derived or [],
            }
            PlayerService.save_stat_definitions(world_name, stat_definitions)
            logger.info(f"Saved {len(validated.stat_system.stats)} stat definitions")

            # Initialize player state
            state = PlayerService.load_player_state(world_name)
            if state:
                # Set default stats
                default_stats = {stat.name: stat.default for stat in validated.stat_system.stats}
                # Apply any overrides
                if validated.initial_stats:
                    default_stats.update(validated.initial_stats)
                state.stats = default_stats
                PlayerService.save_player_state(world_name, state)
                logger.info(f"Initialized player state with {len(state.stats)} stats")

            # Write full lore (overwrites draft, preserving world notes if they exist)
            existing_lore = WorldService.load_lore(world_name)
            world_notes_marker = "\n\n---\n## World Notes"
            if world_notes_marker in existing_lore:
                # Preserve the World Notes section
                notes_index = existing_lore.find(world_notes_marker)
                world_notes_section = existing_lore[notes_index:]
                new_lore = f"# World Lore\n\n{validated.lore}{world_notes_section}"
            else:
                new_lore = f"# World Lore\n\n{validated.lore}"

            # Add world notes if provided
            if validated.world_notes:
                world_notes = validated.world_notes.replace("\\n", "\n")
                new_lore = f"{new_lore}\n\n---\n## World Notes\n{world_notes}"

            WorldService.save_lore(world_name, new_lore)
            logger.info(f"✅ World persisted for '{world_name}'")

            return {
                "content": [
                    {
                        "type": "text",
                        "text": f"World persisted successfully. Stats: {len(validated.stat_system.stats)}, "
                        f"Lore: {len(validated.lore)} characters",
                    }
                ],
            }

        except Exception as e:
            logger.error(f"Failed to persist world: {e}", exc_info=True)
            return {
                "content": [{"type": "text", "text": f"Error persisting world: {e}"}],
                "is_error": True,
            }

    return persist_world

File: backend-ts/src/sdk/handlers/onboarding-tools.ts
import { addCharacterToLocation, getLocationByName } from '../../crud/locations'
import { getAgentByName } from '../../crud/agents'
import type { LocationStorage } from '../../services/location-storage'
import type { PlayerService } from '../../services/player-service'
import type { WorldService } from '../../services/world-service'
import { WorldResetService } from '../../services/world-reset-service'
import { getLoreGuidelinesConfig } from '../loaders/yaml-config'
import { formatTemplate } from '../tools/definitions'
import {
  completeTool,
  draftWorldTool,
  persistWorldTool,
  readLoreGuidelinesTool,
} from '../tools/onboarding'
import { resolveTool } from '../tools/registry'
import { getLogger } from '../../infrastructure/logging/logger'
import type { AgentFilesystemService } from '../../services/agent-filesystem-service'
import { tool, requireWorldName, toolError, toolSuccess, type SdkTool, type ToolContext } from './context'

/**
 * World initialisation. Port of `sdk/handlers/onboarding_tools.py`.
 *
 * The three writing tools run in a fixed order and each leaves the world in a
 * state the next one depends on — see `tools/onboarding.ts` for why. What is
 * worth stating here is what `complete` does *beyond* flipping the phase: it
 * captures the initial-state snapshot that the reset feature restores to, and
 * that snapshot has to be taken after the item designer has run, or "reset this
 * world" hands the player an empty inventory they never started with.
 */

const logger = getLogger('OnboardingTools')

export interface OnboardingDeps {
  worlds: WorldService
  players: PlayerService
  locations: LocationStorage
  reset: WorldResetService
  /** Needed to place onboarding-created NPCs; without it `complete` still works. */
  agentFiles?: AgentFilesystemService
}

/**
 * The lore guidelines text, read from `lore_guidelines.yaml`'s active version.
 *
 * Python routes this through `get_tool_description("lore_guidelines")`, a
 * special case inside the tool-description lookup that has nothing to do with
 * tool descriptions. Read directly here; the indirection carried no behaviour.
 */
function loadLoreGuidelines(): string {
  const config = getLoreGuidelinesConfig()
  const activeVersion = typeof config.active_version === 'string' ? config.active_version : 'v1'
  const version = config[activeVersion]
  if (typeof version !== 'object' || version === null) return ''
  const template = (version as Record<string, unknown>).template
  return typeof template === 'string' ? template : ''
}

/** Marker `persist_world` preserves across a lore rewrite. */
const WORLD_NOTES_MARKER = '\n\n---\n## World Notes'

export function createOnboardingTools(ctx: ToolContext, deps: OnboardingDeps): SdkTool[] {
  const worldName = requireWorldName(ctx)
  const tools: SdkTool[] = []

  // --------------------------------------------------------------------
  // read_lore_guidelines
  // --------------------------------------------------------------------
  const loreDef = resolveTool(readLoreGuidelinesTool.name, ctx.groupName)
  if (loreDef) {
    // Read once, at build time. The file is hot-reloaded through the mtime cache
    // anyway, and a turn that reads it twice must not see two different texts.
    const guidelines = loadLoreGuidelines()
    tools.push(
      tool(
        readLoreGuidelinesTool.name,
        loreDef.description,
        readLoreGuidelinesTool.inputSchema,
        async () => toolSuccess(guidelines),
      ),
    )
  }

  // --------------------------------------------------------------------
  // draft_world
  // --------------------------------------------------------------------
  //
  // Unconditional in Python — `create_onboarding_tools` gates
  // `read_lore_guidelines` and `complete` on `is_tool_enabled` and appends
  // these two without asking. Gated here anyway: an unconditional tool cannot
  // be turned off from a group config, and there is no reason for these two to
  // be the exception.
  const draftDef = resolveTool(draftWorldTool.name, ctx.groupName)
  if (draftDef) {
    tools.push(
      tool(
        draftWorldTool.name,
        draftDef.description,
        draftWorldTool.inputSchema,
        async (args) => {
          try {
            const config = deps.worlds.ensureWorldExists(worldName)
            config.genre = args.genre
            config.theme = args.theme
            deps.worlds.saveWorldConfig(worldName, config)

            // Overwritten wholesale by `persist_world` later; this exists only so
            // the sub-agents running in the background have a world to design for.
            deps.worlds.saveLore(worldName, `# World Lore\n\n${args.lore_summary}`)

            return toolSuccess(
              `World draft created. Genre: ${args.genre}, Theme: ${args.theme}. ` +
                `Sub-agents can now start with this context.`,
            )
          } catch (error) {
            logger.error(`Failed to create world draft: ${String(error)}`)
            return toolError(`Error creating world draft: ${String(error)}`)
          }
        },
      ),
    )
  }

  // --------------------------------------------------------------------
  // persist_world
  // --------------------------------------------------------------------
  const persistDef = resolveTool(persistWorldTool.name, ctx.groupName)
  if (persistDef) {
    tools.push(
      tool(
        persistWorldTool.name,
        persistDef.description,
        persistWorldTool.inputSchema,
        async (args) => {
          try {
            deps.worlds.ensureWorldExists(worldName)

            const stats = args.stat_system.stats
            deps.players.saveStatDefinitions(worldName, {
              stats: stats.map((stat) => ({ ...stat })),
              derived: args.stat_system.derived ?? [],
            })

            const state = deps.players.loadPlayerState(worldName)
            if (state) {
              const values: Record<string, number> = {}
              for (const stat of stats) values[stat.name] = stat.default
              // Overrides layer *over* the declared defaults rather than
              // replacing the map, so a world can start the player at half health
              // without having to restate every other stat.
              Object.assign(values, args.initial_stats ?? {})
              state.stats = values
              deps.players.savePlayerState(worldName, state)
            }

            // The draft is replaced, but anything already filed under
            // "## World Notes" survives — the notes are written by other agents
            // and are not the Onboarding Manager's to discard.
            const existingLore = deps.worlds.loadLore(worldName)
            const notesIndex = existingLore.indexOf(WORLD_NOTES_MARKER)
            let newLore =
              notesIndex === -1
                ? `# World Lore\n\n${args.lore}`
                : `# World Lore\n\n${args.lore}${existingLore.slice(notesIndex)}`

            if (args.world_notes) {
              // The model emits literal backslash-n inside a YAML-ish string often
              // enough that Python un-escapes it by hand before writing.
              const notes = args.world_notes.replaceAll('\\n', '\n')
              newLore = `${newLore}\n\n---\n## World Notes\n${notes}`
            }

            deps.worlds.saveLore(worldName, newLore)

            return toolSuccess(
              `World persisted successfully. Stats: ${stats.length}, ` +
                `Lore: ${args.lore.length} characters`,
            )
          } catch (error) {
            logger.error(`Failed to persist world: ${String(error)}`)
            return toolError(`Error persisting world: ${String(error)}`)
          }
        },
      ),
    )
  }

  // --------------------------------------------------------------------
  // complete
  // --------------------------------------------------------------------
  const completeDef = resolveTool(completeTool.name, ctx.groupName)
  if (completeDef) {
    tools.push(
      tool(
        completeTool.name,
        completeDef.description,
        completeTool.inputSchema,
        async (args) => {
          const {
            player_name: playerName,
            starting_location: startingLocation,
            starting_hour: startingHour,
          } = args

          try {
            // Checked against the *filesystem* index and by exact folder name.
            // A display name here would pass the model's own sense check and then
            // leave the player standing at a location that does not resolve.
            const fsLocations = deps.locations.loadAllLocations(worldName)
            if (!Object.hasOwn(fsLocations, startingLocation)) {
              const available = Object.keys(fsLocations)
              return toolError(
                `Starting location '${startingLocation}' not found. ` +
                  `Available locations: ${available.length > 0 ? available.join(', ') : 'none'}. ` +
                  `Make sure location_designer created this location first.`,
              )
            }

            const config = deps.worlds.ensureWorldExists(worldName)
            config.userName = playerName
            // *Pending*, not applied: the phase flips once this turn ends, so the
            // agent finishing its response is not reading a world that changed
            // underneath it mid-turn.
            config.pendingPhase = 'active'
            deps.worlds.saveWorldConfig(worldName, config)

            const playerState = deps.players.loadPlayerState(worldName)
            if (playerState) {
              playerState.currentLocation = startingLocation
              playerState.gameTime = { hour: startingHour, minute: 0, day: 1 }
              deps.players.savePlayerState(worldName, playerState)

              deps.reset.saveInitialState(
                worldName,
                WorldResetService.createInitialStateSnapshot({
                  startingLocation,
                  initialStats: playerState.stats,
                  initialInventory: playerState.inventory,
                  initialGameTime: playerState.gameTime,
                }),
              )
            }

            if (deps.agentFiles && ctx.worldId !== undefined) {
              try {
                addWorldAgentsToInitialLocation(ctx, deps.agentFiles, worldName, startingLocation)
              } catch (error) {
                logger.warning(`Failed to add agents to initial location: ${String(error)}`)
              }
            }
          } catch (error) {
            logger.error(`Failed to complete onboarding: ${String(error)}`)
            return toolError(`Error completing onboarding: ${String(error)}`)
          }

          return toolSuccess(
            formatTemplate(completeDef.response, {
              player_name: playerName,
              starting_location: startingLocation,
              starting_hour: startingHour,
            }),
          )
        },
      ),
    )
  }

  return tools
}

/**
 * Place every NPC onboarding created at the starting location.
 *
 * They were created against the *onboarding room*, which is an interview and
 * not a place, so until this runs they exist in the world with nowhere to
 * stand. Port of `_add_world_agents_to_initial_location`; per-agent failures
 * are logged and skipped, because one unplaceable NPC must not cost the player
 * every other one.
 */
function addWorldAgentsToInitialLocation(
  ctx: ToolContext,
  agentFiles: AgentFilesystemService,
  worldName: string,
  initialLocationName: string,
): number {
  const db = ctx.getDb()
  const agentNames = agentFiles.listWorldAgents(worldName)
  if (agentNames.length === 0) return 0

  const location = getLocationByName(db, ctx.worldId!, initialLocationName)
  if (!location) {
    logger.warning(`Initial location '${initialLocationName}' not found in database`)
    return 0
  }

  let added = 0
  for (const agentName of agentNames) {
    try {
      // World-unscoped, matching Python's `get_agent_by_name(db, agent_name)`.
      const agent = getAgentByName(db, agentName)
      if (!agent) {
        logger.warning(`Agent '${agentName}' not found in database`)
        continue
      }
      addCharacterToLocation(db, agent.id, location.id)
      added += 1
    } catch (error) {
      logger.warning(`Failed to add agent '${agentName}' to location: ${String(error)}`)
    }
  }

  logger.info(`Added ${added} agents to initial location`)
  return added
}
